[12.0][IMP] sale_order_product_recommendation: Mobile views

pedrobaeza opened the pull request:
Add kanban + form for proper mobile usage.

![Selección_036](https://user-images.githubusercontent.com/7165771/77347496-3e5b7a00-6d38-11ea-82fe-782748df763e.png)

![Selección_037](https://user-images.githubusercontent.com/7165771/77347545-4d422c80-6d38-11ea-9442-7cb02fb25f1f.png)

cc @Tecnativa TT21495

pedrobaeza committed:
[IMP] sale_order_product_recommendation: Mobile views

Add kanban + form for proper mobile usage.

sergio-teruel commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -173,6 +173,7 @@ class SaleOrderRecommendationLine(models.TransientModel):
     sale_line_id = fields.Many2one(
         comodel_name="sale.order.line",
     )
+    sale_uom = fields.Many2one(related="sale_line_id.product_uom")
```
sale_uom or sale_uom_id? I know that the raleted field is only product_uom...

pedrobaeza replied:
Yeah, I follow a bit the related name, but no problem in renaming it

sergio-teruel replied:
Good for me...

yajo replied:
IMHO it's better to rename it.

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation_view.xml:
```diff
@@ -28,6 +29,38 @@
+                                            </div>
+                                            <div class="oe_kanban_details">
+                                                <div><field name="product_id"/></div>
+                                                <div>Price: <field name="price_unit" widget="monetary" options="{'currency_field': 'currency_id', 'field_digits': True, 'style': 'margin-bottom: 0px'}" class="mt-0"/></div>
```
```suggestion
                                                <div>Price: <field name="price_unit" widget="monetary" options="{'currency_field': 'currency_id', 'field_digits': True}" class="mt-0 mb-0"/></div>
```

pedrobaeza replied:
This was several tests @sergio-teruel and me did. Do you think with that it's going to work?

yajo replied:
I think it should work, but if you tested it and it doesn't, this is also not a critical thing... It's just a good rule of thumb to use classes instead of inline styles wherever possible.

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation_view.xml:
```diff
@@ -28,6 +29,38 @@
+                                                <div><field name="product_id"/></div>
+                                                <div>Price: <field name="price_unit" widget="monetary" options="{'currency_field': 'currency_id', 'field_digits': True, 'style': 'margin-bottom: 0px'}" class="mt-0"/></div>
+                                                <div>Delivered: <field name="times_delivered"/> times - <field name="units_delivered"/> <field name="sale_uom"/></div>
+                                                <div><strong><span>To be included: </span><field name="units_included"/><span> </span><field name="sale_uom"/></strong></div>
```
Don't you need this?

```suggestion
                                                <div><strong><span>To be included: </span><field name="units_included" widget="numeric_step"/><span> </span><field name="sale_uom"/></strong></div>
```

pedrobaeza replied:
No, the widget is not working in kanban.

chienandalu reviewed (commented):
Testing in runbot with mobile device (it can be reproduced with browser tools) when you try to add units from the kanban view, they're not added to sale order:
<details>
<img src="https://user-images.githubusercontent.com/5040182/77519547-fc3d5000-6e7f-11ea-83e2-e0ad03103477.png" alt="image" style="max-width:100%;">
<img src="https://user-images.githubusercontent.com/5040182/77519622-15de9780-6e80-11ea-9419-0e1539d1148f.png" alt="image" style="max-width:100%;">
<img src="https://user-images.githubusercontent.com/5040182/77519651-24c54a00-6e80-11ea-9457-aba55f8f72f2.png" alt="image" style="max-width:100%;">
</details>

pedrobaeza committed:
[IMP] sale_order_product_recommendation: Remove useless flag `is_modified`

It's better to simply check the flag condition than to keep that field.

pedrobaeza committed:
[FIX] sale_order_recommendation_secondary_unit: Don't tie second. unit to product default one

Having a related field to default sale secondary UoM makes this recommendator to not fit
possible different secondary UoM to be used.

This commit fixes this behavior. A refactoring on the main module having a prepare
method hook has been needed.

pedrobaeza committed:
[IMP] sale_order_product_recommendation*: Use Form for populating onchanges

Previous code was not very resilient to possible extra adaptations or require
specific code to be added to the old hook `_trigger_so_line_onchanges`.

With this, we make sure all onchanges are executed thanks to the odoo.tests.Form
class.

We lose a bit of performance, but it's better than to have issues with data not
correctly synchronized.

pedrobaeza commented:
The investigation of the problem depicted by @chienandalu has finished in a total revamp of the module for improving several parts. As it has been huge, I have split in several commits self-explaining with the commit message. It can be interesting that you review them separately.

pedrobaeza commented:
Ready to be reviewed, although runbot seems a bit contaminated.

Can you please check @chienandalu and @Yajo ?

chienandalu commented:
>  runbot seems a bit contaminated.
It is. I had to uninstall sale and install the modules again.

When I try to open the wizard I get this exception:

```python
Odoo Server Error
Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 98, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 962, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 759, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 746, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/models.py", line 5533, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/.repo_requirements/odoo/odoo/models.py", line 5377, in _onchange_eval
    method_res = method(self)
  File "/home/odoo/build/OCA/sale-workflow/sale_order_product_recommendation/wizards/sale_order_recommendation.py", line 108, in _generate_recommendations
    self._prepare_recommendation_line_vals(found_line, line)
  File "/home/odoo/build/OCA/sale-workflow/sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py", line 15, in _prepare_recommendation_line_vals
    group_line, so_line=so_line)
  File "/home/odoo/build/OCA/sale-workflow/sale_order_product_recommendation/wizards/sale_order_recommendation.py", line 67, in _prepare_recommendation_line_vals
    "product_id": group_line["product_id"][0],
KeyError: 'product_id'
```

chienandalu commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -87,25 +104,19 @@ def _generate_recommendations(self):
-                "sale_line_id": line.id,
-            })
+            new_line = recommendation_lines.new(
+                self._prepare_recommendation_line_vals(found_line, line)
```
this only will work if the product is already in the `found_dict`. Maybe it's a matter of just getting it in the `_prepare_recommendation_line_vals`

pedrobaeza replied:
Fixed!

chienandalu commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -86,26 +103,22 @@ def _generate_recommendations(self):
-                "units_included": line.product_uom_qty,
-                "sale_line_id": line.id,
+            found_line = found_dict.get(line.product_id.id, {
+                "product_id:" (line.product_id.id, False),
```
```suggestion
                "product_id": (line.product_id.id, False),
```

pedrobaeza replied:
Oh, stupid typo. Fixed.

chienandalu reviewed (approved):
Great :) :+1:

pedrobaeza commented:
/ocabot merge major

pedrobaeza commented:
/ocabot merge major
